From 374f991cee23de642e58666dcc86d5b9b95e4488 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Mon, 3 Sep 2018 18:25:29 +0900
Subject: [PATCH 309/909] drm: rcar-du: Add overlay layer size check

commit 70f9c10b38a33a0972ae34e1a7da9a38cd594839 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

The overlay layer size that exceeds the size of the master layer
(desktop layer) is prohibited by the H/W specification.
Also, when CRTC size is 0 x 0, overlay can not be displayed,
so do not check it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_plane.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_plane.c b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
index 64f4768f23e6..0e881a01450f 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_plane.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_plane.c
@@ -570,6 +570,9 @@ int __rcar_du_plane_atomic_check(struct drm_plane *plane,
 	struct drm_device *dev = plane->dev;
 	struct drm_crtc_state *crtc_state;
 	int ret;
+	struct rcar_du_vsp_plane *rplane = to_rcar_vsp_plane(plane);
+	struct rcar_du_device *rcdu = rplane->vsp->dev;
+	int hdis, vdis;
 
 	if (!state->crtc) {
 		/*
@@ -581,6 +584,20 @@ int __rcar_du_plane_atomic_check(struct drm_plane *plane,
 		return 0;
 	}
 
+	hdis = state->crtc->mode.hdisplay;
+	vdis = state->crtc->mode.vdisplay;
+
+	if ((hdis > 0 && vdis > 0) &&
+	    state->plane->type == DRM_PLANE_TYPE_OVERLAY &&
+	    (((state->crtc_w + state->crtc_x) > hdis) ||
+	    ((state->crtc_h + state->crtc_y) > vdis))) {
+		dev_err(rcdu->dev,
+			"%s: specify (%dx%d) + (%d, %d) < (%dx%d).\n",
+			__func__, state->crtc_w, state->crtc_h, state->crtc_x,
+			state->crtc_y, hdis, vdis);
+		return -EINVAL;
+	}
+
 	crtc_state = drm_atomic_get_crtc_state(state->state, state->crtc);
 	if (IS_ERR(crtc_state))
 		return PTR_ERR(crtc_state);
-- 
2.17.1

